### PR TITLE
Use safe temporary directory on Linux

### PR DESCRIPTION
**Pull request by MaxGaukler**
**Description**
The Linux launcher created a predictable temporary directory in an unsafe fashion.

**Related Issues**
This pull-request solves the most critical part of #1566

**Tasks**
Check all non-standard use-cases on Linux
  - [x] snap package
  - [x] external controllers
  - [x] anything else that uses TMPDIR, WEBOTS_TEMP_PATH etc.

### CODE REVIEW

**Comment by MaxGaukler** [comment]
The remaining tasks should be reviewed by someone with more experience of the internal details and of how to build and package webots; I don't have the necessary knowledge to be sure that it works. The standard case (deb on Ubuntu, just replace the launcher script) works for me.

**Comment by omichel** [comment]
OK, we will review that and check the corresponding boxes.

**Comment by DavidMansolino** [comment]
I checked all the possible issues raised by this change and it seems only this function needs update (because it relies on the fact that the temporary folder include the Webots PID):

https://github.com/cyberbotics/webots/blob/master/src/lib/Controller/api/system.c#L131

**Comment by MaxGaukler** [comment]
Yes, that might be an issue for external controllers because they don't 
receive the WEBOTS_TMP_PATH environment variable.

Another solution on Linux would be to use /run/user/$UID/, which is a 
private temporary directory. However, unlike /tmp, that is a RAM-disk, 
so it could run out of space in extreme cases (huge generated 
textures/PROTOs).

**Comment by omichel** [comment]
This `WEBOTS_PID` is used only in the case of extern controllers and multiple instances of Webots: https://cyberbotics.com/doc/guide/running-extern-robot-controllers#multiple-concurrent-simulations

**Comment by DavidMansolino** [comment]
> Yes, that might be an issue for external controllers because they don't
receive the WEBOTS_TMP_PATH environment variable.

You are completely correct!

> Another solution on Linux would be to use /run/user/$UID/, which is a
private temporary directory. However, unlike /tmp, that is a RAM-disk,
so it could run out of space in extreme cases (huge generated
textures/PROTOs).

I think we can stick with the /tmp solution, we just need to make this function more robust to that it doesn't relies on the pid anymore.
> This `WEBOTS_PID` is used only in the case of extern controllers and multiple instances of Webots: https://cyberbotics.com/doc/guide/running-extern-robot-controllers#multiple-concurrent-simulations

For the `WEBOTS_PID` yes, it is required only in case of multiple instances of Webots, but here the problem is not really `WEBOTS_PID` but rather the fact that the ``wbu_system_webots_tmp_path`` function is expecting a PID (or at least a number) in the temporary folder name (which is not always the case anymore with the `mktemp` function.

**Comment by mgmax** [comment]
You could add some known part to the `mktemp` template, e.g. `mktemp -d /tmp/webots-$PID-XXXXXX`. However, trusting that part may be problematic (?).

**Comment by MaxGaukler** [comment]
You could add some known part to the `mktemp` template, e.g. `mktemp -d /tmp/webots-$PID-XXXXXX`. However, trusting that part may be problematic (?).

**Comment by DavidMansolino** [comment]
> You could add some known part to the `mktemp` template, e.g. `mktemp -d /tmp/webots-$PID-XXXXXX`. However, trusting that part may be problematic (?).

Good to know, this may indeed help a lot. I think about the trust it is quite safe, because at worse, the controller will connect to the wrong instance of Webots but this should not be too critical (as it will not load any library or so from this folder).

File: src/lib/Controller/api/system.c [hunk 1]
```diff
@@ -159,18 +161,39 @@ const char *wbu_system_webots_tmp_path() {
             continue;
           if (s.st_mtime < most_recent)
             continue;
-          sscanf(entry->d_name, "webots-%d", &webots_pid);
+          sscanf(entry->d_name, "webots-%d%s", &webots_pid, random_part);
```
Comment (MaxGaukler) [inline src/lib/Controller/api/system.c:164]: The `random_part` buffer could overflow:
```suggestion
          if (strlen(entry->d_name) > 64) {
              continue; // avoid buffer overflow
          }
          sscanf(entry->d_name, "webots-%d%s", &webots_pid, random_part);
```

Comment (DavidMansolino) [reply src/lib/Controller/api/system.c:164]: Thank you for the pointer, our test suite just warn me about the same issue (https://travis-ci.com/github/cyberbotics/webots/jobs/322059423#L275)

**Comment by DavidMansolino** [comment]
I have checked all the use case (normal controller, extern controller with and without WEBOTS_PID defined) both on Linux and macOS and both are working fine.
Our complete test suite is also working: https://travis-ci.com/github/cyberbotics/webots/builds/161427095

File: src/lib/Controller/api/system.c (hunk 1 above)
Comment (omichel) [reply src/lib/Controller/api/system.c:167]: I would rather use:
```suggestion
          sscanf(entry->d_name, "webots-%d-%63s", &webots_pid, random_part);
```

File: src/lib/Controller/api/system.c [hunk 2]
```diff
@@ -159,18 +161,45 @@ const char *wbu_system_webots_tmp_path() {
...
+          struct stat s;
+          snprintf(path_buffer, path_buffer_size, "%s/%s", tmp, entry->d_name);
+          if (stat(path_buffer, &s) < 0)
+            continue;
+          if (!S_ISDIR(s.st_mode))
+            continue;
+          if (strlen(entry->d_name) > 64)
+            continue;
+          // cppcheck-suppress invalidscanf
+          sscanf(entry->d_name, "webots-%d%s", &webots_pid, random_part);
```
Comment (omichel) [inline src/lib/Controller/api/system.c:191]:
```suggestion
          sscanf(entry->d_name, "webots-%d-%63s", &webots_pid, random_part);
```

File: src/lib/Controller/api/system.c [hunk 3]
```diff
@@ -159,18 +161,45 @@ const char *wbu_system_webots_tmp_path() {
...
+      }
+      closedir(dir);
+    }
   }
   if (webots_pid == 0) {
     free(path_buffer);
     path_buffer = NULL;
   } else
-    snprintf(path_buffer, path_buffer_size, "%s/webots-%d", tmp, webots_pid);
+    snprintf(path_buffer, path_buffer_size, "%s/webots-%d%s", tmp, webots_pid, random_part);
```
Comment (omichel) [inline src/lib/Controller/api/system.c:202]:
```suggestion
    snprintf(path_buffer, path_buffer_size, "%s/webots-%d-%s", tmp, webots_pid, random_part);
```

**Comment by DavidMansolino** [comment]
> I believe it's cleaner to use %63s to avoid the buffer overflow problem rather than adding a test and silence the warning.

This can remove the need to silent the warning indeed but it is dangerous (and therefore the test should remain) as if such folder exist it will only partially read the name and then return a non existing folder name.

> Also, it's more readable to include the dash - in the sscanf / sprintf formula.

This is not possible as the dash is not present on MacOS and Windows.

**Comment by omichel** [comment]
I see, so on Windows and mac OS, the random part is an empty string, this is why we should not hard-code the dash. And yes, I understand we should keep the test.

**Comment by DavidMansolino** [comment]
@MaxGaukler thank you for your nice contribution! I am allowing myself to merge this PR.

**Pull request merged**